File: packages/autonomo-ruby/lib/autonomo/state.rb
# frozen_string_literal: true

module Autonomo
  # User context information
  class UserContext
    attr_accessor :id, :email, :role, :extra

    def initialize(id: nil, email: nil, role: nil, **extra)
      @id = id
      @email = email
      @role = role
      @extra = extra
    end

    def to_h
      result = {}
      result[:id] = @id if @id
      result[:email] = @email if @email
      result[:role] = @role if @role
      result.merge!(@extra)
      result
    end
  end

  # Network request information
  class NetworkRequest
    attr_reader :method, :url, :status, :duration, :error

    def initialize(method:, url:, status: nil, duration: nil, error: nil)
      @method = method
      @url = url
      @status = status
      @duration = duration
      @error = error
    end

    def to_h
      result = { method: @method, url: @url }
      result[:status] = @status if @status
      result[:duration] = @duration if @duration
      result[:error] = @error if @error
      result
    end
  end

  # Complete application state snapshot
  class AppState
    attr_reader :screen, :timestamp, :instance, :user, :elements, :custom_actions,
                :data, :errors, :logs, :render_errors, :network

    def initialize(screen:, timestamp:, instance:, user:, elements:, custom_actions:,
                   data:, errors:, logs:, render_errors:, network:)
      @screen = screen
      @timestamp = timestamp
      @instance = instance
      @user = user
      @elements = elements
      @custom_actions = custom_actions
      @data = data
      @errors = errors
      @logs = logs
      @render_errors = render_errors
      @network = network
    end

    def to_h
      result = {
        screen: @screen,
        timestamp: @timestamp,
        elements: @elements.map(&:to_h),
        customActions: @custom_actions,
        errors: @errors,
        logs: @logs,
        renderErrors: @render_errors
      }
      result[:instance] = @instance.to_h if @instance
      result[:user] = @user.to_h if @user
      result[:data] = @data if @data && !@data.empty?
      result[:network] = @network.map(&:to_h) if @network && !@network.empty?
      result
    end
  end

  # Singleton state manager
  class StateManager
    MAX_ERRORS = 50
    MAX_LOGS = 100
    MAX_NETWORK = 50

    @instance = nil

    def self.instance
      @instance ||= new
    end

    def initialize
      @screen = "unknown"
      @user = nil
      @data = {}
      @errors = []
      @logs = []
      @render_errors = []
      @network = []
      @listeners = []
      @mutex = Monitor.new  # Use Monitor for reentrant locking

      # Forward registry/action changes
      Autonomo.registry.on_change { notify_change }
      Autonomo.custom_actions.on_change { notify_change }
    end

    # Set current screen/route
    def set_screen(screen)
      @mutex.synchronize do
        @screen = screen
        notify_change
      end
    end

    # Get current screen
    def get_screen
      @mutex.synchronize { @screen }
    end

    # Set user context
    def set_user(user)
      @mutex.synchronize do
        @user = user
        notify_change
      end
    end

    # Set application data
    def set_data(data)
      @mutex.synchronize do
        @data = data
        notify_change
      end
    end

    # Merge data into existing
    def merge_data(data)
      @mutex.synchronize do
        @data.merge!(data)
        notify_change
      end
    end

    # Add an error
    def add_error(error)
      @mutex.synchronize do
        @errors << error
        @errors.shift while @errors.size > MAX_ERRORS
        notify_change
      end
    end

    # Add a log entry
    def add_log(log)
      @mutex.synchronize do
        @logs << log
        @logs.shift while @logs.size > MAX_LOGS
      end
    end

    # Add a render error
    def add_render_error(error)
      @mutex.synchronize do
        @render_errors << error
        @render_errors.shift while @render_errors.size > MAX_ERRORS
        notify_change
      end
    end

    # Add a network request
    def add_network_request(request)
      @mutex.synchronize do
        @network << request
        @network.shift while @network.size > MAX_NETWORK
      end
    end

    # Clear errors
    def clear_errors
      @mutex.synchronize do
        @errors.clear
        @render_errors.clear
        notify_change
      end
    end

    # Clear logs
    def clear_logs
      @mutex.synchronize { @logs.clear }
    end

    # Clear network history
    def clear_network
      @mutex.synchronize { @network.clear }
    end

    # Get current state snapshot
    def get_state
      @mutex.synchronize do
        AppState.new(
          screen: @screen,
          timestamp: (Time.now.to_f * 1000).to_i,
          instance: Autonomo.get_instance,
          user: @user,
          elements: Autonomo.registry.get_all,
          custom_actions: Autonomo.custom_actions.list,
          data: @data.empty? ? nil : @data.dup,
          errors: @errors.dup,
          logs: @logs.dup,
          render_errors: @render_errors.dup,
          network: @network.empty? ? nil : @network.dup
        )
      end
    end

    # Subscribe to state changes
    def on_change(&listener)
      @mutex.synchronize { @listeners << listener }
      -> { @mutex.synchronize { @listeners.delete(listener) } }
    end

    # Trigger a state update notification
    def notify_change
      state = get_state
      @listeners.each { |l| l.call(state) }
    end
  end

  # Module-level helpers
  class << self
    def state
      StateManager.instance
    end
  end
end

File: packages/autonomo-swift/Sources/Autonomo/State.swift
import Foundation

// MARK: - User Context

/// User context information
public struct UserContext {
    public var id: String?
    public var email: String?
    public var role: String?
    public var extra: [String: Any]
    
    public init(id: String? = nil, email: String? = nil, role: String? = nil, extra: [String: Any] = [:]) {
        self.id = id
        self.email = email
        self.role = role
        self.extra = extra
    }
    
    public func toDictionary() -> [String: Any] {
        var result: [String: Any] = [:]
        if let id = id { result["id"] = id }
        if let email = email { result["email"] = email }
        if let role = role { result["role"] = role }
        for (key, value) in extra { result[key] = value }
        return result
    }
}

// MARK: - Network Request

/// Network request information
public struct NetworkRequest {
    public let method: String
    public let url: String
    public var status: Int?
    public var duration: Int?
    public var error: String?
    
    public init(method: String, url: String, status: Int? = nil, duration: Int? = nil, error: String? = nil) {
        self.method = method
        self.url = url
        self.status = status
        self.duration = duration
        self.error = error
    }
    
    public func toDictionary() -> [String: Any] {
        var result: [String: Any] = ["method": method, "url": url]
        if let status = status { result["status"] = status }
        if let duration = duration { result["duration"] = duration }
        if let error = error { result["error"] = error }
        return result
    }
}

// MARK: - App State

/// Complete application state snapshot
public struct AppState {
    public let screen: String
    public let timestamp: Int64
    public let instance: InstanceInfo?
    public let user: UserContext?
    public let elements: [ElementInfo]
    public let customActions: [String]
    public let data: [String: Any]?
    public let errors: [String]
    public let logs: [String]
    public let renderErrors: [String]
    public let network: [NetworkRequest]?
    
    public func toDictionary() -> [String: Any] {
        var result: [String: Any] = [
            "screen": screen,
            "timestamp": timestamp,
            "elements": elements.map { $0.toDictionary() },
            "customActions": customActions,
            "errors": errors,
            "logs": logs,
            "renderErrors": renderErrors
        ]
        if let instance = instance { result["instance"] = instance.toDictionary() }
        if let user = user { result["user"] = user.toDictionary() }
        if let data = data, !data.isEmpty { result["data"] = data }
        if let network = network, !network.isEmpty {
            result["network"] = network.map { $0.toDictionary() }
        }
        return result
    }
}

// MARK: - State Manager

/// Singleton state manager
public class StateManager {
    public static let shared = StateManager()
    
    private static let maxErrors = 50
    private static let maxLogs = 100
    private static let maxNetwork = 50
    
    private var _screen = "unknown"
    private var _user: UserContext?
    private var _data: [String: Any] = [:]
    private var _errors: [String] = []
    private var _logs: [String] = []
    private var _renderErrors: [String] = []
    private var _network: [NetworkRequest] = []
    private var listeners: [(AppState) -> Void] = []
    private let lock = NSLock()
    
    private init() {
        // Forward registry/action changes
        ElementRegistry.shared.onChange { [weak self] in self?.notifyChange() }
        CustomActionsRegistry.shared.onChange { [weak self] in self?.notifyChange() }
    }
    
    /// Set current screen/route
    public func setScreen(_ screen: String) {
        lock.lock()
        _screen = screen
        lock.unlock()
        notifyChange()
    }
    
    /// Get current screen
    public func getScreen() -> String {
        lock.lock()
        defer { lock.unlock() }
        return _screen
    }
    
    /// Set user context
    public func setUser(_ user: UserContext?) {
        lock.lock()
        _user = user
        lock.unlock()
        notifyChange()
    }
    
    /// Set application data
    public func setData(_ data: [String: Any]) {
        lock.lock()
        _data = data
        lock.unlock()
        notifyChange()
    }
    
    /// Merge data into existing
    public func mergeData(_ data: [String: Any]) {
        lock.lock()
        for (key, value) in data { _data[key] = value }
        lock.unlock()
        notifyChange()
    }
    
    /// Add an error
    public func addError(_ error: String) {
        lock.lock()
        _errors.append(error)
        while _errors.count > Self.maxErrors { _errors.removeFirst() }
        lock.unlock()
        notifyChange()
    }
    
    /// Add a log entry
    public func addLog(_ log: String) {
        lock.lock()
        _logs.append(log)
        while _logs.count > Self.maxLogs { _logs.removeFirst() }
        lock.unlock()
    }
    
    /// Add a render error
    public func addRenderError(_ error: String) {
        lock.lock()
        _renderErrors.append(error)
        while _renderErrors.count > Self.maxErrors { _renderErrors.removeFirst() }
        lock.unlock()
        notifyChange()
    }
    
    /// Add a network request
    public func addNetworkRequest(_ request: NetworkRequest) {
        lock.lock()
        _network.append(request)
        while _network.count > Self.maxNetwork { _network.removeFirst() }
        lock.unlock()
    }
    
    /// Clear errors
    public func clearErrors() {
        lock.lock()
        _errors.removeAll()
        _renderErrors.removeAll()
        lock.unlock()
        notifyChange()
    }
    
    /// Clear logs
    public func clearLogs() {
        lock.lock()
        _logs.removeAll()
        lock.unlock()
    }
    
    /// Clear network history
    public func clearNetwork() {
        lock.lock()
        _network.removeAll()
        lock.unlock()
    }
    
    /// Get current state snapshot
    public func getState() -> AppState {
        lock.lock()
        defer { lock.unlock() }
        return AppState(
            screen: _screen,
            timestamp: Int64(Date().timeIntervalSince1970 * 1000),
            instance: InstanceManager.shared.getInstance(),
            user: _user,
            elements: ElementRegistry.shared.getAll(),
            customActions: CustomActionsRegistry.shared.list(),
            data: _data.isEmpty ? nil : _data,
            errors: _errors,
            logs: _logs,
            renderErrors: _renderErrors,
            network: _network.isEmpty ? nil : _network
        )
    }
    
    /// Subscribe to state changes
    @discardableResult
    public func onChange(_ listener: @escaping (AppState) -> Void) -> () -> Void {
        lock.lock()
        listeners.append(listener)
        let index = listeners.count - 1
        lock.unlock()
        return { [weak self] in
            self?.lock.lock()
            if let self = self, index < self.listeners.count {
                self.listeners.remove(at: index)
            }
            self?.lock.unlock()
        }
    }
    
    /// Trigger a state update notification
    public func notifyChange() {
        let state = getState()
        lock.lock()
        let currentListeners = listeners
        lock.unlock()
        for listener in currentListeners {
            listener(state)
        }
    }
}

// MARK: - Convenience

/// Global state manager accessor
public var state: StateManager { StateManager.shared }
